File: src/test/java/com/piotrkafel/stripe/ShopOpeningTimeProblemTest.java
package com.piotrkafel.stripe;

import org.junit.jupiter.api.Test;

import static com.piotrkafel.stripe.ShopOpeningTimeProblem.computePenalty;
import static com.piotrkafel.stripe.ShopOpeningTimeProblem.findBestClosingTime;
import static org.junit.jupiter.api.Assertions.assertEquals;

public class ShopOpeningTimeProblemTest {

    @Test
    public void shouldCalculatePenaltyCorrectly() {
        assertEquals(3, computePenalty("Y Y Y N N N N", 0));
        assertEquals(4, computePenalty("Y Y Y N N N N", 7));
        assertEquals(0, computePenalty("Y Y Y N N N N", 3));
        assertEquals(0, computePenalty("", 0));
        assertEquals(1, computePenalty("Y N Y N N N N", 3));
    }

    @Test
    public void shouldCalculateBestClosingTimeCorrectly() {
        assertEquals(3, findBestClosingTime("Y Y Y N N N N"));
        assertEquals(0, findBestClosingTime(""));
        assertEquals(1, findBestClosingTime("Y"));
        assertEquals(0, findBestClosingTime("N N N N"));
        assertEquals(4, findBestClosingTime("Y Y Y Y"));
        assertEquals(5, findBestClosingTime("N Y Y Y Y N N N Y N N Y Y N N N N Y Y N N Y N N N"));
        assertEquals(0, findBestClosingTime("N N N N N Y Y Y N N N N Y Y Y N N N Y N Y Y N Y N"));
        assertEquals(25, findBestClosingTime("Y Y N N N Y Y N Y Y N N N Y Y N N Y Y Y N Y N Y Y"));
    }
}

File: src/test/java/com/piotrkafel/stripe/ShopOpeningTimeProblemTest.kt
package com.piotrkafel.stripe

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

class ShopOpeningTimeProblemTest {
    @Test
    fun shouldCalculatePenaltyCorrectly() {
        Assertions.assertEquals(3, ShopOpeningTimeProblem.computePenalty("Y Y Y N N N N", 0))
        Assertions.assertEquals(4, ShopOpeningTimeProblem.computePenalty("Y Y Y N N N N", 7))
        Assertions.assertEquals(0, ShopOpeningTimeProblem.computePenalty("Y Y Y N N N N", 3))
        Assertions.assertEquals(0, ShopOpeningTimeProblem.computePenalty("", 0))
        Assertions.assertEquals(1, ShopOpeningTimeProblem.computePenalty("Y N Y N N N N", 3))
    }

    @Test
    fun shouldFindBestClosingTimeCorrectly() {
        Assertions.assertEquals(3, ShopOpeningTimeProblem.findBestClosingTime("Y Y Y N N N N"))
        Assertions.assertEquals(0, ShopOpeningTimeProblem.findBestClosingTime(""))
        Assertions.assertEquals(1, ShopOpeningTimeProblem.findBestClosingTime("Y"))
        Assertions.assertEquals(0, ShopOpeningTimeProblem.findBestClosingTime("N N N N"))
        Assertions.assertEquals(4, ShopOpeningTimeProblem.findBestClosingTime("Y Y Y Y"))
        Assertions.assertEquals(5, ShopOpeningTimeProblem.findBestClosingTime("N Y Y Y Y N N N Y N N Y Y N N N N Y Y N N Y N N N"))
        Assertions.assertEquals(0, ShopOpeningTimeProblem.findBestClosingTime("N N N N N Y Y Y N N N N Y Y Y N N N Y N Y Y N Y N"))
        Assertions.assertEquals(25, ShopOpeningTimeProblem.findBestClosingTime("Y Y N N N Y Y N Y Y N N N Y Y N N Y Y Y N Y N Y Y"))
    }

    @Test
    fun shouldGetAllBestClosingTimeByBeginCorrectly() {
        Assertions.assertEquals(listOf(2, 0), ShopOpeningTimeProblem.getAllClosingTimesOrderByBegin("BEGIN Y Y END BEGIN N N END"))
        Assertions.assertEquals(listOf(1, 2, 2), ShopOpeningTimeProblem.getAllClosingTimesOrderByBegin("BEGIN BEGIN BEGIN Y Y N Y END Y Y N N END Y N Y N END"))
        Assertions.assertEquals(listOf(1, 0, 1), ShopOpeningTimeProblem.getAllClosingTimesOrderByBegin("BEGIN Y N BEGIN N N BEGIN Y N END Y END Y END"))
        Assertions.assertEquals(listOf(1, 0), ShopOpeningTimeProblem.getAllClosingTimesOrderByBegin("BEGIN Y N END BEGIN N N Y END"))
    }

    @Test
    fun shouldGetAllBestClosingTimeByEndCorrectly() {
        Assertions.assertEquals(listOf(2, 0), ShopOpeningTimeProblem.getAllClosingTimesOrderByEnd("BEGIN Y Y END BEGIN N N END"))
        Assertions.assertEquals(listOf(2, 2, 1), ShopOpeningTimeProblem.getAllClosingTimesOrderByEnd("BEGIN BEGIN BEGIN Y Y N Y END Y Y N N END Y N Y N END"))
        Assertions.assertEquals(listOf(1, 0, 1), ShopOpeningTimeProblem.getAllClosingTimesOrderByEnd("BEGIN Y N BEGIN N N BEGIN Y N END Y END Y END"))
        Assertions.assertEquals(listOf(1, 0), ShopOpeningTimeProblem.getAllClosingTimesOrderByEnd("BEGIN Y N END BEGIN N N Y END"))
    }
}
